src/Credentials/InstanceProfileProvider.php

```diff
@@ -241,23 +155,136 @@ public function __invoke($previousCredentials = null)
+
+            try {
+                return $this->request($path, 'GET', $headers)->wait();
+            } catch (TransferException $e) {
+                if ($this->apiVersion === self::API_VERSION_EXTENDED
+                    && $this->getExceptionStatusCode($e) === 404
+                ) {
+                    $this->apiVersion = self::API_VERSION_LEGACY;
+                }
+
+                $this->handleRetryableException(
+                    $e,
+                    ['blacklist' => [401, 403]],
+                    $this->createErrorMessage($e->getMessage())
+                );
+            }
+
+            $this->attempts++;
+        }
+    }
+
+    /**
+     * @param array $headers
+     * @param $previousCredentials
+     *
+     * @return mixed|null
+     */
+    private function getCredentials(array $headers, $previousCredentials): array | null
+    {
+        while (true) {
+            $path = $this->getMetadataPath() . $this->profile;
+
+            try {
+                $json = $this->request($path, 'GET', $headers)->wait();
+                return $this->decodeResult($json);
+            } catch (InvalidJsonException $e) {
+                $this->handleRetryableException(
+                    $e,
+                    ['blacklist' => [401, 403]],
+                    $this->createErrorMessage(
+                        'Invalid JSON response, retries exhausted'
+                    )
+                );
+            } catch (TransferException $e) {
+                if ($this->apiVersion === self::API_VERSION_EXTENDED
+                    && $this->getExceptionStatusCode($e) === 404
+                ) {
+                    $this->apiVersion = self::API_VERSION_LEGACY;
+                }
+
+                if ($previousCredentials instanceof Credentials
+                    && ($this->getExceptionStatusCode($e) === 500
+                        || str_contains($e->getMessage(), "cURL error 28"))
+                ) {
+                    return null;
+                }
+
+                $this->handleRetryableException(
+                    $e,
+                    ['blacklist' => [401, 403]],
+                    $this->createErrorMessage($e->getMessage())
+                );
+            }
+
+            $this->attempts++;
+        }
+    }
+
+    /**
+     * @param string $path
      * @param string $method
      * @param array $headers
      * @return PromiseInterface Returns a promise that is fulfilled with the
      *                          body of the response as a string.
      */
-    private function request($url, $method = 'GET', $headers = [])
+    private function request($path, $method = 'GET', $headers = [])
     {
-        $disabled = getenv(self::ENV_DISABLE) ?: false;
-        if (strcasecmp($disabled, 'true') === 0) {
+        $disabled = ConfigurationResolver::ini(self::CFG_DISABLE_EC2_METADATA, 'bool')
+            ?? ConfigurationResolver::env(substr(self::ENV_DISABLE, 4), 'bool')
```

Comment:
nit: Why not just to use getenv instead? and that will not require to normalize self::ENV_DISABLE and will avoid having to do a method call.

Reply:
`getenv()` requires us to do the string parsing and/or conversion to bool, whereas the call to the configuration provider guarantees we either receive a boolean or null.

Reply:
We could just do:
```
$disabled = getenv(self::ENV_DISABLE);
if ($disabled == true) {

}
```

And that will avoid doing any null check or boolean conversion. What do you think?